src/Sdk/Cerbos/Sdk/Cloud/V1/Interceptor/AuthInterceptor.cs

```diff
@@ -0,0 +1,152 @@
+            ClientInterceptorContext<TRequest, TResponse> context,
+            BlockingUnaryCallContinuation<TRequest, TResponse> continuation
+        )
+        {
+            ThrowIfUnauthenticated();
+
+            if (Expired())
+            {
+                IssueToken();
+            }
+
+            var newContext = new ClientInterceptorContext<TRequest, TResponse>(
+                context.Method,
+                context.Host,
+                context.Options.WithHeaders(MetadataWithAuthToken(context.Options.Headers))
+            );
+            return continuation(request, newContext);
+        }
+        
+        public override AsyncUnaryCall<TResponse> AsyncUnaryCall<TRequest, TResponse>(
+            TRequest request,
+            ClientInterceptorContext<TRequest, TResponse> context,
+            AsyncUnaryCallContinuation<TRequest, TResponse> continuation)
+        {
+            ThrowIfUnauthenticated();
+
+            if (Expired())
+            {
+                IssueToken();
+            }
+
+            var newContext = new ClientInterceptorContext<TRequest, TResponse>(
+                context.Method,
+                context.Host,
+                context.Options.WithHeaders(MetadataWithAuthToken(context.Options.Headers))
+            );
+            return continuation(request, newContext);
+        }
+
+        private void IssueToken()
+        {
+            RWLock.AcquireWriterLock(0);
+
+            try
+            {
+                if (!Expired())
+                {
+                    return;
+                }
+
+                var response = ApiKeyClient.IssueAccessToken(Credentials.ToIssueAccessTokenRequest());
+                AccessToken = response.AccessToken;
+                AccessTokenExpiresAt = response.ExpiresAt;
+                LastException = null;
+            }
+            catch (RpcException e)
+            {
+                if (e.StatusCode == StatusCode.Unauthenticated)
+                {
+                    LastException = e;
+                }
+
+                throw;
+            }
+            catch (Exception e)
+            {
+                throw new Exception($"Failed to issue access token: ${e}");
+            }
+            finally
+            {
+                RWLock.ReleaseWriterLock();
+            }
+        }
+
+        private void ThrowIfUnauthenticated()
+        {
+            RWLock.AcquireReaderLock(0);
+            try
+            {
+                if (LastException is RpcException && LastException.StatusCode == StatusCode.Unauthenticated)
+                {
+                    throw LastException;
+                }
+            }
+            finally
+            {
+                RWLock.ReleaseReaderLock();
+            }
+        }
+
+        private bool Expired()
+        {
+            RWLock.AcquireReaderLock(0);
+            try
+            {
+                if (AccessTokenExpiresAt.CompareTo(DateTime.MinValue) == 0)
```

Comment:
I assume this is because an unset `DateTime` object is set to `DateTime.MinValue` instead of null?

Reply:
Yes